llvm/test/tools/llvm-symbolizer/approximate-line-handcrafted.s

```diff
@@ -0,0 +1,261 @@
+# REQUIRES: x86-registered-target
+
+# RUN: clang -O3 -gline-tables-only -T%S/linker-script.ld --target=x86_64-pc-linux %s -o %t.o
```

Comment:
We cannot invoke `clang` or linker in llvm/test tests. 

If this is complex, perhaps `cross-project-tests/debuginfo-tests/` can be used.

Reply:
> If this is complex, perhaps `cross-project-tests/debuginfo-tests/` can be used.

Yes, This test is complex in the sense as we are creating a custom section to have another sequence apart from .text sequence. So as per @dwblaikie request to add a lit test which properly check the sequence boundaries for approximation. To cover that requirement a minimum of two sections has to be created. These two sections(.text & custom section) will have corresponding end_sequence flag setup in the final object file when line table is parsed by the llvm-symbolizer.  we then use llvm-symbolizer binary to check if approximation is enabled then it shouldn't cross sequence boundaries.

Ok, but where to add these in ```cross-project-tests/debuginfo-tests```?
Should I create a new directory, in ```cross-project-tests/debuginfo-tests``` ? 
Is it ok to create a directory ```llvm-symbolizer-tests```  in ```cross-project-tests/debuginfo-tests``` ?

If added into ```cross-project-tests/debuginfo-tests``` then how upstream CI would execute a check for this?

Reply:
After reading from this https://llvm.org/docs/TestingGuide.html#debugging-information-tests about ```cross-project-tests/debuginfo-tests```. I thing tests under directory ```debuginfo-tests``` are meant to be execute under a debugger. 

This particular lit test case don't require to execute under a debugger. So,  I think  it should be in ```cross-project-tests```. But how to ascertain that this lit test should always be run for llvm-symbolizer to which I am not sure. Any comments?

Reply:
I hope that we don't need compilation or linking. I would try hand-writing some asm to compile to an object file (via llvm-mc) with two sequences, with the second sequence having higher addresses than the first. I think this would allow testing the "sequence crossing disallowed" condition by using one of the higher addresses that wouldn't appear in the first sequence. Perhaps we could combine both test cases with this approach?

Reply:
> I hope that we don't need compilation or linking. 
I am also not in favour of compilation or linking but to cover that sequence crossing should not happen we need linker to invoked.

Let me explain this in detail.
The assembly(containing two sequences) whether handwritten or whether generated has addresses of sequences mapped at offset zero always by llvm-mc. 

I checked the address of sequences by running below commands on approximate-line-handcrafted.s(Note No .loc entries are edited here.)
 
```llvm-mc -g -filetype=obj -triple=x86_64-pc-linux approximate-line-handcrafted.s -o approximate-line-handcrafted.o``` 

```llvm-dwarfdump --debug-line approximate-line-handcrafted.o``` outputs this line table:-

```
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
**0x0000000000000000**      1     78      1   0             0       0  is_stmt prologue_end
0x0000000000000006      1     78      1   0             0       0  is_stmt end_sequence
**0x0000000000000000**      2     49      1   0             0       0  is_stmt prologue_end
0x0000000000000003      2     39      1   0             0       0
0x0000000000000010      3      0      1   0             0       0  is_stmt
0x0000000000000012      3     49      1   0             0       0  is_stmt prologue_end
0x0000000000000014      3     39      1   0             0       0
0x0000000000000020      4      0      0   0             0       0  is_stmt
0x0000000000000021      7      2      0   0             0       0  is_stmt prologue_end
0x0000000000000034      8      2      0   0             0       0  is_stmt
0x0000000000000047      9      2      0   0             0       0  is_stmt
0x000000000000005a     10      3      0   0             0       0  is_stmt
0x000000000000005c     10      3      0   0             0       0  epilogue_begin
0x000000000000005e     10      3      0   0             0       0  end_sequence
```
Now in above o/p you see that two sequences have there offset at 0x0000000000000000. When I try to run llvm-symbolizer on those address I get weird behaviour like below.

```
~$ llvm-symbolizer --obj=approximate-line-handcrafted.o 0x04
.L.str
/tmp/test/./definitions.h:2:39

~$ llvm-symbolizer --obj=approximate-line-handcrafted.o 0x1
.L.str
/tmp/test/./definitions.h:2:49
```

This happens when we have two sequences in object file but their mapping are offset at 0x00. In approximate-line-generated.s this was not happening because it has single sequence. If multiple sequences are there in object then llvm-mc dosen't offset the sections as no memory map is happening. This job is for the linker. So, to cover this part I had to invoke the linker with the dummy linker script(which is static for all targets). The linker script provides memory mapping by offsetting sections at a particular offset(Here in approximate-line-handcrafted.s at  0x500000).

So with this below command:-
```
clang -O3 -T linker-script.ld -gline-tables-only --target=x86_64-pc-linux approximate-line-handcrafted.s -o approximate-line-handcrafted.o

llvm-dwarfdump --debug-line approximate-line-handcrafted.o

**Output Line table**
Address            Line   Column File   ISA Discriminator OpIndex Flags
------------------ ------ ------ ------ --- ------------- ------- -------------
0x0000000000500790      1     78      1   0             0       0  is_stmt prologue_end
0x0000000000500796      1     78      1   0             0       0  is_stmt end_sequence
0x00000000005000f0      2     49      1   0             0       0  is_stmt prologue_end
0x00000000005000f3      2     39      1   0             0       0
0x0000000000500100      3      0      1   0             0       0  is_stmt
0x0000000000500102      3     49      1   0             0       0  is_stmt prologue_end
0x0000000000500104      3     39      1   0             0       0
0x0000000000500110      4      0      0   0             0       0  is_stmt
0x0000000000500111      7      2      0   0             0       0  is_stmt prologue_end
0x0000000000500124      8      2      0   0             0       0  is_stmt
0x0000000000500137      9      2      0   0             0       0  is_stmt
0x000000000050014a     10      3      0   0             0       0  is_stmt
0x000000000050014c     10      3      0   0             0       0  epilogue_begin
0x000000000050014e     10      3      0   0             0       0  end_sequence

~$ llvm-symbolizer --obj=approximate-line-handcrafted.o 0x5000f3
add
/tmp/test/./definitions.h:2:39

~$ llvm-symbolizer --obj=approximate-line-handcrafted.o 0x500790
dummy_function
/tmp/test/./definitions.h:1:78

```   

Which is great. I can easily modify here some of the line table entries with value zero here and invoke llvm-symbolizer on unique addresses to cover that sequence boundaries should not happen.   

All I am saying is invoking ```llvm-mc``` dosen't work with multiple sequences as default offset for every sequence starts at 0x00. If there is a possibility of any instruction  x86_64 which can set offset correctly for sequences(I didn't find any), then I could make this work wihout hassle of calling linker.

Reply:
I've not really had the time to review this, as I have too much else on my plate, but have you considered whether yaml2obj would achieve your goal for hand crafting a .debug_line section? There are some examples under llvm/test/yaml2obj/ELF/DWARF/debug-line.yaml of how the format works.

Reply:
Sequences don't have to reset the address to 0. I'll post an example..

Reply:
I started with some compiler output for two.c:
```
int foo()
{
return 10;
}

int bar()
{
foo();
return 20;
}
```

Then I cleaned up the asm and added a handwritten .debug_line section:

```
	.text
	.section	.text.foo,"ax",@progbits
	.hidden	foo
	.globl	foo
	.type	foo,@function
foo:
.Lfunc_begin0:
	movl	$10, %eax
	retq
.Ltmp0:
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo

	.hidden	bar
	.globl	bar
	.type	bar,@function
bar:
.Lfunc_begin1:
# %bb.0:
	pushq	%rbp
	movq	%rsp, %rbp
.Ltmp1:
	callq	foo
	movl	$20, %eax
	popq	%rbp
	retq
.Ltmp2:
.Lfunc_end1:
	.size	bar, .Lfunc_end1-bar

	.section	.debug_abbrev,"",@progbits
	.byte	1                               # Abbreviation Code
	.byte	17                              # DW_TAG_compile_unit
	.byte	0                               # DW_CHILDREN_no
	.byte	37                              # DW_AT_producer
	.byte	14                              # DW_FORM_strp
	.byte	19                              # DW_AT_language
	.byte	5                               # DW_FORM_data2
	.byte	3                               # DW_AT_name
	.byte	14                              # DW_FORM_strp
	.byte	16                              # DW_AT_stmt_list
	.byte	23                              # DW_FORM_sec_offset
	.byte	27                              # DW_AT_comp_dir
	.byte	14                              # DW_FORM_strp
	.byte	83                              # DW_AT_use_UTF8
	.byte	25                              # DW_FORM_flag_present
	.byte	17                              # DW_AT_low_pc
	.byte	1                               # DW_FORM_addr
	.byte	85                              # DW_AT_ranges
	.byte	23                              # DW_FORM_sec_offset
	.byte	0                               # EOM(1)
	.byte	0                               # EOM(2)
	.byte	0                               # EOM(3)
	.section	.debug_info,"",@progbits
.Lcu_begin0:
	.long	.Ldebug_info_end0-.Ldebug_info_start0 # Length of Unit
.Ldebug_info_start0:
	.short	4                               # DWARF version number
	.long	.debug_abbrev                   # Offset Into Abbrev. Section
	.byte	8                               # Address Size (in bytes)
	.byte	1                               # Abbrev [1] 0xb:0x1f DW_TAG_compile_unit
	.long	.Linfo_string0                  # DW_AT_producer
	.short	29                              # DW_AT_language
	.long	.Linfo_string1                  # DW_AT_name
	.long	.Lline_table_start0             # DW_AT_stmt_list
	.long	.Linfo_string2                  # DW_AT_comp_dir
                                        # DW_AT_use_UTF8
	.quad	0                               # DW_AT_low_pc
	.long	.Ldebug_ranges0                 # DW_AT_ranges
.Ldebug_info_end0:
	.section	.debug_aranges,"",@progbits
	.section	.debug_ranges,"",@progbits
.Ldebug_ranges0:
	.quad	.Lfunc_begin0
	.quad	.Lfunc_end0
	.quad	.Lfunc_begin1
	.quad	.Lfunc_end1
	.quad	0
	.quad	0
	.section	.debug_str,"MS",@progbits,1
.Linfo_string0:
	.asciz	"clang version 16.0.5 ---------------------------------------" # string offset=0
.Linfo_string1:
	.asciz	"two.c"                         # string offset=61
.Linfo_string2:
	.asciz	"c:\\Temp\\dwarfline"           # string offset=67
	.ident	"clang version 16.0.5 ---------------------------------------"
	.section	".note.GNU-stack","",@progbits
.function_and_data_sections
	.section	.debug_line,"",@progbits
.Lline_table_start0:
        .long .Lunit_end - .Lunit_start # unit length
.Lunit_start:
        .short 4   # version
        .long .Lprologue_end - .Lprologue_start # header length
.Lprologue_start:
	.byte 1                                # minimum_instruction_length
	.byte 1                                # maximum_operations_per_instruction
	.byte 0                                # default_is_stmt
	.byte -5                               # line_base
	.byte 14                               # line_range
	.byte 13                               # opcode_base
	.byte 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1 # arguments in standard opcodes
	.asciz "dir0"                          # include directory
	.asciz "dir1"                          # include directory
	.byte 0                                # end of include directories
	.asciz "two.c"                         # filename
	.byte 0                                # reference to dir0
	.byte 0                                # modification time
	.byte 0                                # length of file (unavailable)
	.byte 0                                # end of filenames
.Lprologue_end:
        .byte 0, 9, 2        # DW_LNE_set_address
        .quad 0x0            #  baz
        .byte 3              # DW_LNS_advance_line
        .sleb128 9           #  by 9 (to 10)
        .byte 1              # DW_LNS_copy
        .byte 3              # DW_LNS_advance_line
        .sleb128 -1          #  by -1 (to 9)
        .byte 2              # DW_LNS_advance_pc
        .byte 1              #  += (1 * min instruction length)
        .byte 1              # DW_LNS_copy
        .byte 3              # DW_LNS_advance_line
        .sleb128 -1          #  by -1 (to 8)
        .byte 2              # DW_LNS_advance_pc
        .byte 5              #  += (5 * min instruction length)
        .byte 0, 1, 1        # DW_LNE_end_sequence
        .byte 0, 9, 2        # DW_LNE_set_address
        .quad .Lfunc_begin1 - .Lfunc_begin0 #  wibble
        .byte 3              # DW_LNS_advance_line
        .sleb128 -1           #  by -1 (to 0)
        .byte 1              # DW_LNS_copy
        .byte 2              # DW_LNS_advance_pc
        .byte 1              #  += (1 * min instruction length)
        .byte 1              # DW_LNS_copy
        .byte 3              # DW_LNS_advance_line
        .sleb128 5           #  by 5 (to 5)
        .byte 2              # DW_LNS_advance_pc
        .byte 15             #  += (15 * min instruction length)
        .byte 0, 1, 1        # DW_LNE_end_sequence
.Lunit_end:
```

Hopefully, that's enough to understand the idea. My reading of DWARF suggests that this is legal as there's nothing to disallow two contiguous sequences AFAICS. This might be worth checking: @pogo59 or @dwblaikie?

Note: It may be better to try yaml2obj as suggested by James.

Reply:
Yes, Thanks @jh7370 & @bd1976bris  for pointing the yaml2obj utility. I was not aware that we can represent object file in terms of yaml.